crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs

```diff
@@ -953,4 +953,140 @@ impl ReachabilityConstraints {
             }
         }
     }
+
+    /// Narrow a type based on the predicates in a reachability constraint.
+    ///
+    /// When code is only reachable under certain conditions (encoded in the reachability constraint),
+    /// those conditions can be used to narrow types. For example, if code is only reachable when
+    /// `val is not None`, we can narrow `val` from `int | None` to `int`.
+    ///
+    /// This function traverses the reachability constraint TDD and extracts predicates that can
+    /// narrow the given place, then applies them to the base type.
+    ///
+    /// The `binding` parameter is used to filter out predicates that were created before the
+    /// binding. Such predicates are about an earlier binding of the same place and shouldn't
+    /// be used to narrow the current binding.
+    pub(crate) fn narrow_by_reachability<'db>(
+        &self,
+        db: &'db dyn Db,
+        predicates: &Predicates<'db>,
+        id: ScopedReachabilityConstraintId,
+        place: crate::semantic_index::place::ScopedPlaceId,
+        binding: crate::semantic_index::definition::Definition<'db>,
+        base_ty: Type<'db>,
+    ) -> Type<'db> {
+        // Get the binding's NodeIndex. We'll use this to filter out predicates that were
+        // created before this binding (and thus are about an earlier binding of the same place).
+        let binding_index = binding.kind(db).target_node_index();
```

Comment:
I'm pretty pessimistic that any approach relying on node-index ordering here is going to work out well, since source order does not always correspond to control flow. It can work for simple `if` statements...

Reply:
I'm off the charts not happy with this code :-) Once I realized _something_ was necessary to change how narrowing intersected with re-binding, this was one of the dumber things that could work. But it's why I left this in draft, because I'm not happy with this as landable.

Which is all by way of saying: I appreciate the pointer to the past comment, will take a look!

Reply:
:-) I'm not sure where this fits in your own prioritization, but I'm likely to look at this over the holidays.

Reply:
This is pretty high priority for us; I would have otherwise hoped to work on it in the next month or so. Not opposed to you giving it a try if you want, happy to offer some guidance as needed/useful.